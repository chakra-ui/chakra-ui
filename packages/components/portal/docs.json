{
  "PortalManager": {
    "zIndex": {
      "type": "number",
      "required": false,
      "description": "[Z-Index war] If your has multiple elements\nwith z-index clashing, you might need to apply a z-index to the Portal manager"
    }
  },
  "Portal": {
    "appendToParentPortal": {
      "type": "boolean",
      "defaultValue": true,
      "required": false,
      "description": "If `true`, the portal will check if it is within a parent portal\nand append itself to the parent's portal node.\nThis provides nesting for portals.\n\nIf `false`, the portal will always append to `document.body`\nregardless of nesting. It is used to opt out of portal nesting."
    },
    "containerRef": {
      "type": "RefObject<HTMLElement | null>",
      "required": false,
      "description": "The `ref` to the component where the portal will be attached to."
    }
  }
}
