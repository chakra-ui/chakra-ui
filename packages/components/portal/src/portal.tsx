import { useSafeLayoutEffect } from "@chakra-ui/react-use-safe-layout-effect"
import { createContext } from "@chakra-ui/react-context"
import { createPortal } from "react-dom"
import { usePortalManager } from "./portal-manager"
import { useEffect, useMemo, useRef, useState } from "react"

type PortalContext = HTMLDivElement | null

const [PortalContextProvider, usePortalContext] = createContext<PortalContext>({
  strict: false,
  name: "PortalContext",
})

const PORTAL_CLASSNAME = "chakra-portal"
const PORTAL_SELECTOR = `.chakra-portal`

const Container = (props: React.PropsWithChildren<{ zIndex: number }>) => (
  <div
    className="chakra-portal-zIndex"
    style={{
      position: "absolute",
      zIndex: props.zIndex,
      top: 0,
      left: 0,
      right: 0,
      // NB: Don't add `bottom: 0`, it makes the entire app unusable
      // @see https://github.com/chakra-ui/chakra-ui/issues/3201
    }}
  >
    {props.children}
  </div>
)

/**
 * Portal that uses `document.body` as container
 */
const DefaultPortal = (
  props: React.PropsWithChildren<{ appendToParentPortal?: boolean }>,
) => {
  const { appendToParentPortal, children } = props

  const [tempNode, setTempNode] = useState<HTMLElement | null>(null)
  const portal = useRef<HTMLDivElement | null>(null)

  const [, forceUpdate] = useState({})
  useEffect(() => forceUpdate({}), [])

  const parentPortal = usePortalContext()
  const manager = usePortalManager()

  useSafeLayoutEffect(() => {
    if (!tempNode) return

    const doc = tempNode.ownerDocument
    const host = appendToParentPortal ? parentPortal ?? doc.body : doc.body

    if (!host) return

    portal.current = doc.createElement("div")
    portal.current.className = PORTAL_CLASSNAME

    host.appendChild(portal.current)
    forceUpdate({})

    const portalNode = portal.current
    return () => {
      if (host.contains(portalNode)) {
        host.removeChild(portalNode)
      }
    }
  }, [tempNode])

  const _children = manager?.zIndex ? (
    <Container zIndex={manager?.zIndex}>{children}</Container>
  ) : (
    children
  )

  return portal.current ? (
    createPortal(
      <PortalContextProvider value={portal.current}>
        {_children}
      </PortalContextProvider>,
      portal.current,
    )
  ) : (
    <span
      ref={(el) => {
        if (el) setTempNode(el)
      }}
    />
  )
}

interface ContainerPortalProps extends React.PropsWithChildren<{}> {
  containerRef: React.RefObject<HTMLElement | null>
  /**
   * @default false
   */
  appendToParentPortal?: boolean
}

/**
 * Portal that uses a custom container
 */
const ContainerPortal = (props: ContainerPortalProps) => {
  const { children, containerRef, appendToParentPortal } = props
  const containerEl = containerRef.current
  const host =
    containerEl ?? (typeof window !== "undefined" ? document.body : undefined)

  const portal = useMemo(() => {
    const node = containerEl?.ownerDocument.createElement("div")
    if (node) node.className = PORTAL_CLASSNAME
    return node
  }, [containerEl])

  const [, forceUpdate] = useState({})
  useSafeLayoutEffect(() => forceUpdate({}), [])

  useSafeLayoutEffect(() => {
    if (!portal || !host) return
    host.appendChild(portal)
    return () => {
      host.removeChild(portal)
    }
  }, [portal, host])

  if (host && portal) {
    return createPortal(
      <PortalContextProvider value={appendToParentPortal ? portal : null}>
        {children}
      </PortalContextProvider>,
      portal,
    )
  }

  return null
}

export interface PortalProps {
  /**
   * The `ref` to the component where the portal will be attached to.
   */
  containerRef?: React.RefObject<HTMLElement | null>
  /**
   * The content or node you'll like to portal
   */
  children: React.ReactNode
  /**
   * If `true`, the portal will check if it is within a parent portal
   * and append itself to the parent's portal node.
   * This provides nesting for portals.
   *
   * If `false`, the portal will always append to `document.body`
   * regardless of nesting. It is used to opt out of portal nesting.
   *
   * @default true
   */
  appendToParentPortal?: boolean
}

/**
 * Portal
 *
 * Declarative component used to render children into a DOM node
 * that exists outside the DOM hierarchy of the parent component.
 *
 * @see Docs https://chakra-ui.com/portal
 */

export function Portal(props: PortalProps) {
  const portalProps: PortalProps = {
    appendToParentPortal: true,
    ...props,
  }

  const { containerRef, ...rest } = portalProps
  return containerRef ? (
    <ContainerPortal containerRef={containerRef} {...rest} />
  ) : (
    <DefaultPortal {...rest} />
  )
}

Portal.className = PORTAL_CLASSNAME
Portal.selector = PORTAL_SELECTOR

Portal.displayName = "Portal"
