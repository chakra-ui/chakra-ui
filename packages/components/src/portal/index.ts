export { PortalManager, usePortalManager } from "./portal-manager"
export type { PortalManagerProps } from "./portal-manager"
export { Portal } from "./portal"
export type { PortalProps } from "./portal"
