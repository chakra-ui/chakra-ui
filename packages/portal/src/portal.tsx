import { useForceUpdate, useSafeLayoutEffect } from "@chakra-ui/hooks"
import { isBrowser, __DEV__ } from "@chakra-ui/utils"
import { createContext } from "@chakra-ui/react-utils"
import * as React from "react"
import { createPortal } from "react-dom"
import { usePortalManager } from "./portal-manager"

type PortalContext = HTMLDivElement | null

const [PortalContextProvider, usePortalContext] = createContext<PortalContext>({
  strict: false,
  name: "PortalContext",
})

const PORTAL_CLASSNAME = "chakra-portal"
const PORTAL_SELECTOR = `.chakra-portal`

const Container: React.FC<{ zIndex: number }> = (props) => (
  <div
    className="chakra-portal-zIndex"
    style={{
      position: "absolute",
      zIndex: props.zIndex,
      top: 0,
      left: 0,
      right: 0,
      // NB: Don't add `bottom: 0`, it makes the entire app unusable
      // @see https://github.com/chakra-ui/chakra-ui/issues/3201
    }}
  >
    {props.children}
  </div>
)

/**
 * Portal that uses `document.body` as container
 */
const DefaultPortal: React.FC<{ appendToParentPortal?: boolean }> = (props) => {
  const { appendToParentPortal, children } = props

  const tempNode = React.useRef<HTMLDivElement | null>(null)
  const portal = React.useRef<HTMLDivElement | null>(null)

  const forceUpdate = useForceUpdate()

  const parentPortal = usePortalContext()
  const manager = usePortalManager()

  useSafeLayoutEffect(() => {
    if (!tempNode.current) return

    const doc = tempNode.current!.ownerDocument
    const host = appendToParentPortal ? parentPortal ?? doc.body : doc.body

    if (!host) return

    portal.current = doc.createElement("div")
    portal.current.className = PORTAL_CLASSNAME

    host.appendChild(portal.current)
    forceUpdate()

    const portalNode = portal.current
    return () => {
      if (host.contains(portalNode)) {
        host.removeChild(portalNode)
      }
    }
  }, [])

  const _children = manager?.zIndex ? (
    <Container zIndex={manager?.zIndex}>{children}</Container>
  ) : (
    children
  )

  return portal.current ? (
    createPortal(
      <PortalContextProvider value={portal.current}>
        {_children}
      </PortalContextProvider>,
      portal.current,
    )
  ) : (
    <span ref={tempNode} />
  )
}

interface ContainerPortalProps {
  containerRef: React.RefObject<HTMLElement | null>
  appendToParentPortal?: boolean
}

/**
 * Portal that uses a custom container
 */
const ContainerPortal: React.FC<ContainerPortalProps> = (props) => {
  const { children, containerRef, appendToParentPortal } = props
  const containerEl = containerRef.current
  const host = containerEl ?? (isBrowser ? document.body : undefined)

  const portal = React.useMemo(() => {
    const node = containerEl?.ownerDocument.createElement("div")
    if (node) node.className = PORTAL_CLASSNAME
    return node
  }, [containerEl])

  const forceUpdate = useForceUpdate()

  useSafeLayoutEffect(() => {
    forceUpdate()
  }, [])

  useSafeLayoutEffect(() => {
    if (!portal || !host) return
    host.appendChild(portal)
    return () => {
      host.removeChild(portal)
    }
  }, [portal, host])

  if (host && portal) {
    return createPortal(
      <PortalContextProvider value={appendToParentPortal ? portal : null}>
        {children}
      </PortalContextProvider>,
      portal,
    )
  }

  return null
}

export interface PortalProps {
  /**
   * The `ref` to the component where the portal will be attached to.
   */
  containerRef?: React.RefObject<HTMLElement | null>
  /**
   * The content or node you'll like to portal
   */
  children: React.ReactNode
  /**
   * If `true`, the portal will check if it is within a parent portal
   * and append itself to the parent's portal node.
   * This provides nesting for portals.
   *
   * If `false`, the portal will always append to `document.body`
   * regardless of nesting. It is used to opt out of portal nesting.
   */
  appendToParentPortal?: boolean
}

/**
 * Portal
 *
 * Declarative component used to render children into a DOM node
 * that exists outside the DOM hierarchy of the parent component.
 *
 * @see Docs https://chakra-ui.com/portal
 */

export function Portal(props: PortalProps) {
  const { containerRef, ...rest } = props
  return containerRef ? (
    <ContainerPortal containerRef={containerRef} {...rest} />
  ) : (
    <DefaultPortal {...rest} />
  )
}

Portal.defaultProps = {
  appendToParentPortal: true,
}

Portal.className = PORTAL_CLASSNAME
Portal.selector = PORTAL_SELECTOR

if (__DEV__) {
  Portal.displayName = "Portal"
}
