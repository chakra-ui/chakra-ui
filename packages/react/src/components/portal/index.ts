export { Portal } from "./portal"
export type { PortalProps } from "./portal"
