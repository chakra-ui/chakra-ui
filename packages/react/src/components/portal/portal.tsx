import { createPortal } from "react-dom"

export interface PortalProps {
  /**
   * The `ref` to the component where the portal will be attached to.
   */
  containerRef?: React.RefObject<HTMLElement | null>
  /**
   * The content or node you'll like to portal
   */
  children: React.ReactNode
}

/**
 * Portal
 *
 * Declarative component used to render children into a DOM node
 * that exists outside the DOM hierarchy of the parent component.
 *
 * @see Docs https://chakra-ui.com/portal
 */

export function Portal(props: PortalProps) {
  const { containerRef, children } = props
  const mountNode = containerRef?.current ?? globalThis?.document.body
  return mountNode ? createPortal(children, mountNode) : null
}
