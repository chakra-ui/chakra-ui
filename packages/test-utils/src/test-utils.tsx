import theme from "@chakra-ui/theme"
import { GlobalStyle, ThemeProvider } from "@chakra-ui/system"
import CSSReset from "@chakra-ui/css-reset"
import "@testing-library/jest-dom/extend-expect"
import { render, RenderOptions, fireEvent } from "@testing-library/react"
import * as React from "react"
import { toHaveNoViolations } from "jest-axe"
import serializer from "jest-emotion"

expect.addSnapshotSerializer(serializer)

expect.extend(toHaveNoViolations)

const AllProviders = ({ children }: { children?: React.ReactNode }) => (
  <ThemeProvider theme={theme}>
    <CSSReset />
    <GlobalStyle />
    {children}
  </ThemeProvider>
)

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllProviders, ...options })

export * from "@testing-library/react"

export {
  act as invoke,
  renderHook,
  RenderHookOptions,
  RenderHookResult,
} from "@testing-library/react-hooks"

export { default as userEvent } from "@testing-library/user-event"

export { customRender as render }

export * from "jest-axe"

export const escape = (ui: HTMLElement) =>
  fireEvent.keyDown(ui, { key: "Escape", keyCode: 27 })
