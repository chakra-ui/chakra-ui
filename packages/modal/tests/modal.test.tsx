import * as React from "react"
import { render, fireEvent, axe, press, wait } from "@chakra-ui/test-utils"
import { PortalManager } from "@chakra-ui/portal"
import {
  Modal,
  ModalBody,
  ModalCloseButton,
  ModalContent,
  ModalFooter,
  ModalHeader,
  ModalOverlay,
} from "../src"

const renderWithPortal = (ui: React.ReactElement) =>
  render(<PortalManager>{ui}</PortalManager>)

test("should render correctly", () => {
  const tools = renderWithPortal(
    <Modal isOpen onClose={jest.fn()}>
      <ModalOverlay>
        <ModalContent>
          <ModalHeader>Modal header</ModalHeader>
          <ModalCloseButton />
          <ModalBody>Modal body</ModalBody>
          <ModalFooter>Modal footer</ModalFooter>
        </ModalContent>
      </ModalOverlay>
    </Modal>,
  )
  expect(tools.asFragment()).toMatchSnapshot()
})

test("should have no accessibility violations", async () => {
  const tools = renderWithPortal(
    <Modal isOpen onClose={jest.fn()}>
      <ModalOverlay>
        <ModalContent>
          <ModalHeader>Modal header</ModalHeader>
          <ModalCloseButton />
          <ModalBody>Modal body</ModalBody>
          <ModalFooter>Modal footer</ModalFooter>
        </ModalContent>
      </ModalOverlay>
    </Modal>,
  )

  const result = await axe(tools.container)
  expect(result).toHaveNoViolations()
})

test("should have the proper 'aria' attributes", () => {
  const tools = renderWithPortal(
    <Modal isOpen onClose={jest.fn()}>
      <ModalOverlay>
        <ModalContent data-testid="modal">
          <ModalHeader>Modal header</ModalHeader>
          <ModalBody>Modal body</ModalBody>
        </ModalContent>
      </ModalOverlay>
    </Modal>,
  )

  const dialog = tools.getByTestId("modal")

  /**
   * should have `aria-modal` set to `true`
   */
  expect(dialog).toHaveAttribute("aria-modal", "true")
  expect(dialog).toHaveAttribute("role", "dialog")

  /**
   * The id of `DialogBody` should equal the `aria-describedby` of the dialog
   */
  expect(tools.getByText("Modal body").id).toEqual(
    dialog.getAttribute("aria-describedby"),
  )

  /**
   * The id of `DialogHeader` should equal the `aria-labelledby` of the dialog
   */
  expect(tools.getByText("Modal header").id).toEqual(
    dialog.getAttribute("aria-labelledby"),
  )
})

test("should fire 'onClose' callback when close button is clicked", () => {
  const onClose = jest.fn()

  const tools = renderWithPortal(
    <Modal isOpen onClose={onClose}>
      <ModalOverlay>
        <ModalContent>
          <ModalHeader>Modal header</ModalHeader>
          <ModalCloseButton data-testid="close" />
        </ModalContent>
      </ModalOverlay>
    </Modal>,
  )

  /**
   * click the close button
   */
  fireEvent.click(tools.getByTestId("close"))

  expect(onClose).toHaveBeenCalled()
})

test('clicking overlay or pressing "esc" calls the onClose callback', () => {
  const onClose = jest.fn()
  const tools = renderWithPortal(
    <Modal isOpen onClose={onClose}>
      <ModalOverlay data-testid="overlay">
        <ModalContent>
          <ModalHeader>Modal header</ModalHeader>
          <ModalBody>Modal body</ModalBody>
        </ModalContent>
      </ModalOverlay>
    </Modal>,
  )

  const overlay = tools.getByTestId("overlay")

  // Not working in tests but working in browser. I'm guessing this is due to focus-trap
  // fireEvent.click(overlay as HTMLElement)
  // expect(onClose).toHaveBeenCalled()

  press.Escape(overlay as HTMLElement)
  expect(onClose).toHaveBeenCalled()
})

test("focuses the initial focus ref when opened", () => {
  const Component = () => {
    const [isOpen, setIsOpen] = React.useState(false)
    const inputRef = React.useRef(null)
    return (
      <>
        <button data-testid="button" onClick={() => setIsOpen(true)}>
          Open
        </button>
        <Modal isOpen={isOpen} initialFocusRef={inputRef} onClose={jest.fn()}>
          <ModalOverlay>
            <ModalContent>
              <ModalHeader>Modal header</ModalHeader>
              <ModalBody>
                <input />
                <input />
                <input data-testid="input" ref={inputRef} />
              </ModalBody>
            </ModalContent>
          </ModalOverlay>
        </Modal>
      </>
    )
  }
  const tools = renderWithPortal(<Component />)

  /**
   * User clicks button to open the modal
   */
  fireEvent.click(tools.getByTestId("button"))

  /**
   * We focus the input right away!
   */
  expect(tools.getByTestId("input")).toHaveFocus()
})

test("should return focus to button when closed", async () => {
  const Component = () => {
    const [isOpen, setIsOpen] = React.useState(false)
    const buttonRef = React.useRef(null)
    return (
      <>
        <button
          ref={buttonRef}
          data-testid="button"
          onClick={() => setIsOpen(true)}
        >
          Open
        </button>
        <Modal
          finalFocusRef={buttonRef}
          isOpen={isOpen}
          onClose={() => setIsOpen(false)}
        >
          <ModalOverlay>
            <ModalContent>
              <ModalHeader>Modal header</ModalHeader>
              <ModalCloseButton data-testid="close" />
              <ModalBody>Modal body</ModalBody>
            </ModalContent>
          </ModalOverlay>
        </Modal>
      </>
    )
  }
  const tools = renderWithPortal(<Component />)
  const button = tools.getByTestId("button")

  // make sure button isn't focused at the start
  expect(button).not.toHaveFocus()

  // open and close the modal
  fireEvent.click(button)
  fireEvent.click(tools.getByTestId("close"))

  // wait for button to be focused
  await wait(() => expect(button).toHaveFocus())
})
