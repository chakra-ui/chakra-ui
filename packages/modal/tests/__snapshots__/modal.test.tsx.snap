// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<DocumentFragment>
  .emotion-6 {
  background: rgba(0,0,0,0.48);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  overflow: auto;
  width: 100vw;
  height: 100vh;
  position: fixed;
  left: 0;
  top: 0;
}

.emotion-6.overlay-enter {
  opacity: 0.01;
}

.emotion-6.overlay-enter-active {
  opacity: 1;
  -webkit-transition-timing-function: cubic-bezier(0,0,.2,1);
  transition-timing-function: cubic-bezier(0,0,.2,1);
  -webkit-transition-property: opacity;
  transition-property: opacity;
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
}

.emotion-6.overlay-exit {
  opacity: 1;
}

.emotion-6.overlay-exit-active {
  opacity: 0.01;
  -webkit-transition-timing-function: cubic-bezier(.4,0,1,1);
  transition-timing-function: cubic-bezier(.4,0,1,1);
  -webkit-transition-property: opacity;
  transition-property: opacity;
  -webkit-transition-duration: 100ms;
  transition-duration: 100ms;
}

.emotion-5 {
  border-radius: 0.25rem;
  background: #FFFFFF;
  color: inherit;
  margin-top: 3.75rem;
  margin-bottom: 3.75rem;
  box-shadow: 0 10px 15px -3px rgba(0,0,0,0.1),0 4px 6px -2px rgba(0,0,0,0.05);
  max-width: 28rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  width: 100%;
  outline: 0;
}

.emotion-5.content-enter,
.emotion-5.content-appear {
  opacity: 0.01;
  -webkit-transform: scale(0.97);
  -ms-transform: scale(0.97);
  transform: scale(0.97);
}

.emotion-5.content-enter-active,
.emotion-5.content-appear-active {
  opacity: 1;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  -webkit-transition-timing-function: cubic-bezier(0,0,.2,1);
  transition-timing-function: cubic-bezier(0,0,.2,1);
  -webkit-transition-property: opacity,-webkit-transform;
  -webkit-transition-property: opacity,transform;
  transition-property: opacity,transform;
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
}

.emotion-5.content-exit {
  opacity: 1;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

.emotion-5.content-exit-active {
  opacity: 0.01;
  -webkit-transform: scale(0.97);
  -ms-transform: scale(0.97);
  transform: scale(0.97);
  -webkit-transition-timing-function: cubic-bezier(.4,0,1,1);
  transition-timing-function: cubic-bezier(.4,0,1,1);
  -webkit-transition-property: opacity,-webkit-transform;
  -webkit-transition-property: opacity,transform;
  transition-property: opacity,transform;
  -webkit-transition-duration: 100ms;
  transition-duration: 100ms;
}

.emotion-0 {
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  padding-left: 1.5rem;
  padding-right: 1.5rem;
  padding-top: 1rem;
  padding-bottom: 1rem;
  font-size: 1.25rem;
  font-weight: 600;
}

.emotion-2 {
  border-radius: 0.25rem;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  width: 32px;
  height: 32px;
  outline: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: absolute;
  top: 8px;
  right: 12px;
}

.emotion-2[disabled],
.emotion-2[disabled]:focus,
.emotion-2[disabled]:hover,
.emotion-2[aria-disabled=true],
.emotion-2[aria-disabled=true]:focus,
.emotion-2[aria-disabled=true]:hover,
.emotion-2[data-disabled],
.emotion-2[data-disabled]:focus,
.emotion-2[data-disabled]:hover {
  opacity: 0.4;
  cursor: not-allowed;
  box-shadow: none;
}

.emotion-2:hover,
.emotion-2[data-hover] {
  background: rgba(0,0,0,0.06);
}

.emotion-2:active,
.emotion-2[data-active] {
  background: rgba(0,0,0,0.08);
}

.emotion-2:focus,
.emotion-2[data-focus] {
  box-shadow: 0 0 0 3px rgba(66,153,225,0.6);
}

.emotion-1 {
  font-size: 12px;
  vertical-align: middle;
  display: inline-block;
  line-height: 1em;
  color: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 1em;
  height: 1em;
}

.emotion-3 {
  padding-left: 1.5rem;
  padding-right: 1.5rem;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  padding-left: 1.5rem;
  padding-right: 1.5rem;
  padding-top: 1rem;
  padding-bottom: 1rem;
}

<div
    class="chakra-portal-manager"
  >
    <div
      class="chakra-portal"
    >
      <div
        aria-hidden="true"
        data-aria-hidden="true"
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
      <div
        aria-hidden="true"
        data-aria-hidden="true"
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="1"
      />
      <div
        data-focus-lock-disabled="false"
      >
        <div>
          <div
            class="chakra-modal__overlay emotion-6"
          >
            <section
              aria-describedby="chakra-modal--body-1"
              aria-labelledby="chakra-modal--header-1"
              aria-modal="true"
              class="chakra-modal__content emotion-5 content-appear content-appear-active"
              id="chakra-modal-1"
              role="dialog"
              tabindex="-1"
            >
              <header
                class="chakra-modal__header emotion-0"
                id="chakra-modal--header-1"
              >
                Modal header
              </header>
              <button
                aria-label="Close"
                class="chakra-modal__close-btn emotion-2"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="chakra-icon emotion-1"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
                    fill="currentColor"
                  />
                </svg>
              </button>
              <div
                class="chakra-modal__body emotion-3"
                id="chakra-modal--body-1"
              >
                Modal body
              </div>
              <footer
                class="chakra-modal__footer emotion-4"
              >
                Modal footer
              </footer>
            </section>
          </div>
        </div>
      </div>
      <div
        aria-hidden="true"
        data-aria-hidden="true"
        data-focus-guard="true"
        style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
        tabindex="0"
      />
    </div>
  </div>
</DocumentFragment>
`;
